File: CSV2HTML_by_Java/csv2html/Downloader.java
package csv2html;

import java.awt.image.BufferedImage;
import java.io.File;
import java.util.List;
import utility.ImageUtility;

/**
 * ダウンローダ：CSVファイル・画像ファイル・サムネイル画像ファイルをダウンロードする。
 */
public class Downloader extends IO
{
	/**
	 * ダウンローダのコンストラクタ。
	 * @param aTable テーブル
	 */
	public Downloader(Table aTable)
	{
		super(aTable);

		return;
	}

	/**
	 * 総理大臣の情報を記したCSVファイルをダウンロードする。
	 */
	public void downloadCSV()
	{
		// CSVファイルのURLを取得
		String urlString = this.attributes().csvUrl();

		// URLからテキスト（行リスト）を読み込む
		List<String> lines = IO.readTextFromURL(urlString);

		// ローカルのCSVファイルパスを作成
		String baseDirectory = this.attributes().baseDirectory();
		String csvFileName = "data.csv";
		String csvFilePath = baseDirectory + csvFileName;

		// ファイルに書き出す
		IO.writeText(lines, csvFilePath);

		System.out.println("Downloaded CSV: " + urlString + " -> " + csvFilePath);

		return;
	}

	/**
	 * 総理大臣の画像群をダウンロードする。
	 */
	public void downloadImages()
	{
		int indexOfImage = this.attributes().indexOfImage();
		this.downloadPictures(indexOfImage);

		return;
	}

	/**
	 * 総理大臣の画像群またはサムネイル画像群をダウンロードする。
	 * @param indexOfPicture 画像のインデックス
	 */
	private void downloadPictures(int indexOfPicture)
	{
		return;
	}

	/**
	 * 総理大臣の画像群をダウンロードする。
	 */
	public void downloadThumbnails()
	{
		int indexOfThumbnail = this.attributes().indexOfThumbnail();
		this.downloadPictures(indexOfThumbnail);

		return;
	}

	/**
	 * 総理大臣の情報を記したCSVファイルをダウンロードして、画像群やサムネイル画像群もダウロードする。
	 */
	public void perform()
	{
		// CSVファイルをダウンロード
		this.downloadCSV();

		// 画像とサムネイルのダウンロード（Phase 4で実装予定）
		// this.downloadImages();
		// this.downloadThumbnails();

		return;
	}
}

File: CSV2HTML_by_Python/csv2html/downloader.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""CSV2HTML：総理大臣と徳川幕府の情報「CSV」からWebページ「HTML」を生成。"""
__author__ = 'AOKI Atsushi'
__version__ = '1.0.7'
__date__ = '2021/01/10 (Created: 2016/01/01)'

import os
# import shutil
import urllib.request

from csv2html.io import IO
# from csv2html.reader import Reader

class Downloader(IO):
	"""ダウンローダ：CSVファイル・画像ファイル・サムネイル画像ファイルをダウンロードする。"""

	def __init__(self, input_table):
		"""ダウンローダのコンストラクタ。"""

		super().__init__(input_table)
		(lambda x: x)(input_table) # NOP

	def download_csv(self):
		"""情報を記したCSVファイルをダウンロードする。"""

		# CSVファイルのURLと保存先のパスを取得
		csv_url = self.attributes().csv_url()
		base_directory = self.attributes().base_directory()
		filename = os.path.basename(csv_url)
		save_path = os.path.join(base_directory, 'data.csv')
		
		# ダウンロードを実行
		try:
			urllib.request.urlretrieve(csv_url, save_path)
			print(f"Downloaded CSV: {csv_url} -> {save_path}")
		except Exception as e:
			print(f"Error downloading CSV: {e}")

	def download_images(self, image_filenames):
		"""画像ファイル群または縮小画像ファイル群をダウンロードする。"""

		base_url = self.attributes().base_url()
		base_directory = self.attributes().base_directory()
		
		for image_filename in image_filenames:
			image_url = base_url + image_filename
			save_path = os.path.join(base_directory, image_filename)
			
			# ディレクトリが存在しない場合は作成
			os.makedirs(os.path.dirname(save_path), exist_ok=True)
			
			try:
				urllib.request.urlretrieve(image_url, save_path)
				# print(f"Downloaded image: {image_url} -> {save_path}")
			except Exception as e:
				print(f"Error downloading image {image_url}: {e}")

	def perform(self):
		"""すべて（情報を記したCSVファイル・画像ファイル群・縮小画像ファイル群）をダウンロードする。"""

		# CSVファイルをダウンロード
		self.download_csv()
		
		# 画像とサムネイルをダウンロード
		input_table = self.table()
		# 画像ファイルのリストを取得してダウンロード
		# self.download_images(input_table.image_filenames())
		# self.download_images(input_table.thumbnail_filenames())
		# self.download_images(input_table.thumbnail_filenames())
		pass
